rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/CompactView.java

```diff
@@ -166,269 +172,107 @@ private void changeSelection(String name) {
         _doNotScroll = false;
     }
 
-    private boolean usedBy(NodeSource ns, String username) {
-        for (String hostid : ns.getHosts().keySet()) {
-            Host host = ns.getHosts().get(hostid);
-            if (usedBy(host, username))
-                return true;
-        }
-        return false;
-    }
-
-    private boolean usedBy(Host h, String username) {
-        for (String nodeid : h.getNodes().keySet()) {
-            Node node = h.getNodes().get(nodeid);
-            if (usedBy(node, username))
-                return true;
+    @Override
+    public void updateByDelta(List<NodeSource> nodeSources, List<Node> nodes) {
+        /* first call : create the components */
+        if (this.normalFlow == null) {
+            initializePanel();
         }
-        return false;
-    }
+        processNodeSources(normalFlow, nodeSources);
 
-    private boolean usedBy(Node n, String username) {
-        return username != null ? username.equals(n.getNodeOwner()) : false;
-    }
+        processNodes(normalFlow, nodes);
 
-    /*
-     * create a copy of the original map containing only resources
-     * used by the current user
-     */
-    private Map<String, NodeSource> filterMyNodes(Map<String, NodeSource> origNodes) {
+        processNodeSources(myNodesFlow, nodeSources);
 
         String username = LoginModel.getInstance().getLogin();
 
-        Map<String, NodeSource> myNodesNsMap = new HashMap<String, NodeSource>();
-        for (String nsid : origNodes.keySet()) {
-            NodeSource origNs = origNodes.get(nsid);
-            NodeSource myNodesNs = null;
-            if (!usedBy(origNs, username)) {
-                continue;
-            } else {
-                myNodesNs = new NodeSource(origNs);
-                myNodesNsMap.put(nsid, myNodesNs);
-            }
+        processNodes(myNodesFlow, nodes.stream().filter(node -> usedBy(node, username)).collect(Collectors.toList()));
 
-            // at this point, at least one node of this ns is being used
-
-            Iterator<String> hostsIt = myNodesNs.getHosts().keySet().iterator();
-            while (hostsIt.hasNext()) {
-                String hostid = hostsIt.next();
-                Host host = myNodesNs.getHosts().get(hostid);
-                if (!usedBy(host, username)) {
-                    hostsIt.remove();
-                    continue;
-                }
+        //         if myNodesFlow already have some nodes, but they do not became to us anymore, then delete them
+        nodes.stream()
+             .filter(node -> !usedBy(node, username))
+             .filter(node -> myNodesFlow.isNodeDrawn(node))
+             .forEach(node -> myNodesFlow.remove(node));
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)